### PR TITLE
Scala 3 compatible tailrec

### PR DESCRIPTION
**Pull request by andreaTP**
References #29929 

This encoding of `tailrec` methods is compatible with Scala 3 checks (thanks for the explanation @smarter !).
To the best of my knowledge, this is not going to degrade runtime perf.

### CODE REVIEW

File: akka-actor/src/main/scala/akka/actor/LightArrayRevolverScheduler.scala [hunk 1]
```diff
@@ -128,15 +128,19 @@ class LightArrayRevolverScheduler(config: Config, log: LoggingAdapter, threadFac
         }
       }
 
-      @tailrec final def cancel(): Boolean = {
+      @tailrec private final def tailrecCancel(): Boolean = {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/LightArrayRevolverScheduler.scala:131]: could this def be defined inside the `def cancel()`?

Comment (patriknw) [reply akka-actor/src/main/scala/akka/actor/LightArrayRevolverScheduler.scala:131]: what is the requirement? does it have to be a `private def`? strange  that it wouldn't  allow `final def`

**Comment by smarter** [comment]
Note that you don't need `@tailrec` on a method for tailrec calls to be optimized, so you could also just remove `@tailrec` in situation where you have both tailrec and non-tailrec calls.

File: akka-actor/src/main/scala/akka/actor/LightArrayRevolverScheduler.scala (hunk 1 above)
Comment (andreaTP) [reply akka-actor/src/main/scala/akka/actor/LightArrayRevolverScheduler.scala]: The requirement is that the `@tailrec` shouldn't occur on a method overridden from the super-class.
Declared all of them inside the `def` body, Scala 3 compiler accept even that encoding. 👍

**Comment by andreaTP** [comment]
@smarter I think that the signature here is to prevent a contribution from accidentally making those sensitive cycles not optimized.

**Review by johanandren — APPROVED** [review]

**Review by raboof — APPROVED** [review]

**Comment by raboof** [comment]
OK TO TEST

**Comment by patriknw** [comment]
Superseded by https://github.com/akka/akka/pull/29956, I assume

**Pull request closed without merging**